rfcs/0140-simple-package-paths.md

```diff
@@ -0,0 +1,397 @@
+| --- | --- |
+| start-date | 2022-09-02 |
+| author | Silvan Mosberger (@infinisil) |
+| co-authors | Robert Hensing (@roberth) |
+| pre-RFC reviewers | Thomas Bereknyei (@tomberek), John Ericson (@Ericson2314), Alex Ameen (@aakropotkin) |
+| shepherd-team | @phaer @06kellyjac @aakropotkin @piegamesde |
+| shepherd-leader | - |
+| related-issues | https://github.com/NixOS/nixpkgs/pull/211832 |
+
+# Summary
+[summary]: #summary
+
+Auto-generate trivial top-level attribute definitions in Nixpkgs' `pkgs/top-level/all-packages.nix` from a Nixpkgs-internally standardised directory structure that matches the attribute name.
+This makes it much easier to contribute new packages, since there's no more guessing needed as to where the package should go, both in the ad-hoc directory categories and in `all-packages.nix`.
+
+# Motivation
+[motivation]: #motivation
+
+- It is not obvious to package contributors where to add files or which ones to edit. These are very common questions:
+  - Which directory should my package definition go in?
+  - What are all the categories and do they matter?
+  - What if the package has multiple matching categories?
+  - Why can't I build my package after adding the package file?
+  - Where in `all-packages.nix` should my package go?
+- Figuring out where an attribute is defined is a bit tricky:
+  - First one has to find the definition of it in `all-packages.nix` to see what file it refers to
+    - On GitHub this is even more problematic, as the `all-packages.nix` file is [too big to be displayed by GitHub](https://github.com/NixOS/nixpkgs/blob/nixos-22.05/pkgs/top-level/all-packages.nix)
+  - Then go to that file's definition, which takes quite some time for navigation (unless you have a plugin that can jump to it directly)
+    - It also slows down or even locks up editors due to the file size
+  - `nix edit -f . package-attr` works, though that's not yet stable (it relies on the `nix-command` feature being enabled) and doesn't work with packages that don't set `meta.position` correctly).
+- `all-packages.nix` frequently causes merge conflicts. It's a point of contention for all new packages
+
+# Detailed design
+[design]: #detailed-design
+
+This RFC consists of two parts, each of which is implemented with a PR to Nixpkgs.
+These PR's should be done after a release to maximize the testing period and minimize merge conflicts.
+
+## PR 1: The unit directory standard
+
+This part establishes the new _unit directory standard_ in Nixpkgs.
+This standard is internal to Nixpkgs and not exposed as public interface.
+This standard must be documented in the Nixpkgs manual.
+This PR will be backported to the stable release in order to ensure that backports of new packages work.
+
+### The name "unit"
+
+The term "unit" is chosen partly because it's not associated with any pre-existing assumptions about what it means, which should cause people unfamiliar with this standard to read the documentation.
```

Comment:
This is false, we already have a different meaning of «unit»: https://nixos.org/manual/nixos/stable/index.html#sec-unit-handling

Reply:
Systemd has one, and so do the sciences, and mathematics has a couple kinds of "units". Organizations have units... They're all different.

It's a very broad term, so I wouldn't say that using it for something else would be unexpected. That said, we should weaken the language here a bit.

```suggestion
The term "unit" is chosen partly because it's generally not associated with any pre-existing assumptions about what it means, which should cause people unfamiliar with this standard to read the documentation.
```